tools/syz-testbed/testbed.go

```diff
@@ -82,150 +78,220 @@ func main() {
 		}
-		for _, checkout := range checkouts {
-			fileName := fmt.Sprintf("avg_%v.txt", checkout.Name)
-			saveAvgBenchFile(checkout, filepath.Join(benchFolder, fileName))
+	}()
+
+	ctx.Loop(shutdown)
+}
+
+func (ctx *TestbedContext) GetStatViews() ([]StatView, error) {
+	groupsCompleted := []RunResultGroup{}
+	groupsAll := []RunResultGroup{}
+	for _, checkout := range ctx.Checkouts {
+		running := []*RunResult{}
+		for _, instance := range checkout.Running {
+			result, err := instance.FetchResult()
+			if err != nil {
+				return nil, err
+			}
+			running = append(running, result)
 		}
+		groupsCompleted = append(groupsCompleted, RunResultGroup{
+			Name:    checkout.Name,
+			Results: checkout.Completed,
+		})
+		groupsAll = append(groupsAll, RunResultGroup{
+			Name:    checkout.Name,
+			Results: append(checkout.Completed, append(checkout.Incomplete, running...)...),
+		})
 	}
+	return []StatView{
+		{
+			Name:   "all",
+			Groups: groupsAll,
+		},
+		{
+			Name:   "completed",
+			Groups: groupsCompleted,
+		},
+	}, nil
 }
 
-func runInstance(info InstanceInfo) {
-	logfile, err := os.Create(info.LogFile)
+func (ctx *TestbedContext) saveStatView(view StatView) error {
+	dir := filepath.Join(ctx.Config.Workdir, "stats_"+view.Name)
+	benchDir := filepath.Join(dir, "benches")
+	err := osutil.MkdirAll(benchDir)
 	if err != nil {
-		tool.Failf("[%s] failed to create logfile: %s", info.Name, err)
+		return fmt.Errorf("failed to create %s: %s", benchDir, err)
 	}
-	cmd := osutil.GraciousCommand(info.ExecCommand, info.ExecCommandArgs...)
-	cmd.Stdout = logfile
-	cmd.Stderr = logfile
-	err = cmd.Start()
-	if err != nil {
-		tool.Failf("[%s] failed to start instance: %s", info.Name, err)
+
+	tableStats := map[string]func(view StatView) ([][]string, error){
+		"bugs.csv":           (StatView).GenerateBugTable,
+		"checkout_stats.csv": (StatView).StatsTable,
+		"instance_stats.csv": (StatView).InstanceStatsTable,
+	}
+	for fileName, genFunc := range tableStats {
+		table, err := genFunc(view)
+		if err == nil {
+			SaveTableAsCsv(table, filepath.Join(dir, fileName))
+		} else {
+			log.Printf("some error: %s", err)
+		}
 	}
-	log.Printf("[%s] Instance started. Listening on %s", info.Name, info.HTTP)
-	logfile.Close()
-	err = cmd.Wait()
-	tool.Failf("[%s] Instance exited: %s", info.Name, err)
+	_, err = view.SaveAvgBenches(benchDir)
+	return err
 }
 
-func newCheckout(co TestbedCheckout, cfg *TestbedConfig, managerCfg *mgrconfig.Config) *CheckoutInfo {
-	log.Printf("[%s] Checking out", co.Name)
-	path := filepath.Join(cfg.Workdir, "checkouts", co.Name)
-	if osutil.IsExist(path) {
-		if !*flagCleanup {
-			tool.Failf("path %s already exists", path)
+func (ctx *TestbedContext) TestbedStatsTable() [][]string {
+	table := [][]string{
+		{"Checkout", "Running", "Completed", "Incomplete", "Until reset"},
+	}
+	for _, checkout := range ctx.Checkouts {
+		until := "-"
+		if ctx.NextRestart.After(time.Now()) {
+			until = time.Until(ctx.NextRestart).String()
```

Comment:
This prints time down to nanoseconds. Use Truncate method to truncate to minutes.

Reply:
Added the truncation to seconds, I think they're important, too.